### ISSUE TITLE
Bad Request (400) - Version may not be negative at line 1

### ISSUE THREAD

**Issue by westnordost** [openstreetmap/openstreetmap-website#2240]
This may as well be a bug in StreetComplete or osmapi, I am currently investigating. Maybe you immediately know what is wrong or can help me find the bug.

Since this morning, I am getting error reports from users of StreetComplete about failing upload. The error returned by the OSM API  `Bad Request (400) - Version may not be negative`.

The newest version of the app is out for some weeks already, but I am only getting error reports about this now. So, _something_ must have changed very recently that leads to the app making a bad request.

I can't find where in the OSM API a string like _"Version may not be negative"_ is returned. Could you point me at it? Also, the version mentioned in that error message is the version of the element uploaded, right?

**Comment by westnordost** [comment]
Nevermind, I found the bug. The error message indeed means the version of the uploaded element. StreetComplete sets the version to -1 if no `version` attribute is sent when querying the elements from Overpass.

**Comment by tomhughes** [comment]
If you're using diff upload then those are now going to cgimap which may be generating different errors though I would hope the rails version would also reject a bogus version.

**Comment by westnordost** [comment]
I'm using `POST changeset/$id/upload` with the body being `osmChange`, that's diff upload right?

Was this changed today or in the last few days, the move to cgimap? That would explain why I get those reports now.

If that is the case, I have a hunch why this worked before at all: Maybe the previous rails implementation returned simply a `409 Conflict` if the version given in the diff is not the same as the current server version? Cause that case is covered within the app - the app handles a conflict exception by re-downloading the newest version from the OSM Api and then trying again to upload the element with the changes applied.

**Comment by westnordost** [comment]
Reopened because if my assumption is correct, perhaps it should be considered to return a conflict error in that case so that the cgimap implementation behaves the same as the rails implementation here.

**Comment by tomhughes** [comment]
Yes the switch happened yesterday and yes that is the call that is affected.

@mmd-osm any thoughts?

**Comment by tomhughes** [comment]
I'll move this ticket anyway because if it's not the rails code then this is the wrong place for it...
